keccak256/src/gates/absorb.rs

```diff
@@ -93,67 +93,71 @@ impl<F: FieldExt> AbsorbConfig<F> {
     /// Doc this
     pub fn copy_state_flag_next_inputs(
         &self,
-        region: &mut Region<'_, F>,
-        mut offset: usize,
+        layouter: &mut impl Layouter<F>,
         in_state: [(Cell, F); 25],
         out_state: [F; 25],
         next_input: [F; ABSORB_NEXT_INPUTS],
         flag: (Cell, F),
     ) -> Result<([(Cell, F); 25], (Cell, F)), Error> {
-        let mut state_array = [F::zero(); 25];
-
-        // State at offset + 0
-        for (idx, (cell, value)) in in_state.iter().enumerate() {
-            // Copy value into state_array
-            state_array[idx] = *value;
-            let new_cell = region.assign_advice(
-                || format!("assign state {}", idx),
-                self.state[idx],
-                offset,
-                || Ok(*value),
-            )?;
-
-            region.constrain_equal(*cell, new_cell)?;
-        }
-
-        offset += 1;
-        // Enable `q_mixing` at `offset + 1`
-        self.q_mixing.enable(region, offset)?;
-        // Assign next_mixing at offset + 1
-        for (idx, lane) in next_input.iter().enumerate() {
-            region.assign_advice(
-                || format!("assign next_input {}", idx),
-                self.state[idx],
-                offset,
-                || Ok(*lane),
-            )?;
-        }
-        // Assign flag at last column(17th) of the offset + 1 row.
-        let obtained_cell = region.assign_advice(
-            || format!("assign next_input {}", ABSORB_NEXT_INPUTS),
-            self.state[ABSORB_NEXT_INPUTS],
-            offset,
-            || Ok(flag.1),
-        )?;
-        region.constrain_equal(flag.0, obtained_cell)?;
-
-        offset += 1;
-        // Assign out_state at offset + 2
-        let mut state: Vec<(Cell, F)> = Vec::with_capacity(25);
-        for (idx, lane) in out_state.iter().enumerate() {
-            let cell = region.assign_advice(
-                || format!("assign state {}", idx),
-                self.state[idx],
-                offset,
-                || Ok(*lane),
-            )?;
-            state.push((cell, *lane));
-        }
-        let out_state: [(Cell, F); 25] = state
-            .try_into()
-            .expect("Unexpected into_slice conversion err");
+        layouter.assign_region(
+            || "Absorb state assignations",
+            |mut region| {
+                let mut offset = 0;
+                let mut state_array = [F::zero(); 25];
+                // State at offset + 0
+                for (idx, (cell, value)) in in_state.iter().enumerate() {
+                    // Copy value into state_array
+                    state_array[idx] = *value;
```

Comment:
state_array is assigned but unused. Maybe we should remove it?

Reply:
Good catch!!! Thanks!